apps/vscode-extension/src/intellisense/configuration.ts

```diff
@@ -0,0 +1,62 @@
+import { CompletionItem, CompletionItemKind, CompletionItemProvider, SnippetString } from 'vscode';
+
+interface ConfigurationTags {
+  description: string;
+  detail: string;
+  snippet: string;
+}
+
+const configurationTags: Record<string, ConfigurationTags> = {
+  o3rWidget: {
+    description: 'Tag to use CMS widget for configuration',
+    detail: 'widgetName',
+    snippet: '${1:widgetName}'
+  },
+  o3rWidgetParam: {
+    description: 'Tag to use CMS widget parameter for configuration',
+    detail: 'paramName paramValue',
+    snippet: '${1:paramName} ${2:paramValue}'
+  }
+};
+
+export const configurationCompletionItemProvider: CompletionItemProvider<CompletionItem> = {
+  provideCompletionItems: (doc, pos) => {
+    const completions = Object.entries(configurationTags).map(([label, { detail, description, snippet }]) => {
+      const completion = new CompletionItem({ label: `${label} `, detail }, CompletionItemKind.Keyword);
+      completion.documentation = description;
+      completion.insertText = new SnippetString(`${label} ${snippet}`);
+
+      return completion;
+    });
```

Comment:
this doesn't seem to rely on the function parameters and could be evaluated only once (moving the code outside the function or using a memoize)
if you need to keep it there, at least you should move the code at the very end of the function, to be evaluated only when it's needed